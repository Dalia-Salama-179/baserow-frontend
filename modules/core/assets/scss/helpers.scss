.clearfix::after {
  clear: both;
  content: '';
  display: table;
}

.hidden {
  display: none;
}

.visibility-hidden {
  visibility: hidden;
  transform: translateX(-10000px) translateY(-10000px);
}

.align-right {
  text-align: right;
}

.background-white {
  background-color: $white;
}

.color-primary {
  color: $color-primary-500 !important;
}

.color-success {
  color: $color-success-500 !important;
}

.color-warning {
  color: $color-warning-500 !important;
}

.color-error {
  color: $color-error-500 !important;
}

.margin-top-0 {
  margin-top: 0 !important;
}

.margin-top-1 {
  margin-top: 8px !important;
}

.margin-top-2 {
  margin-top: 16px !important;
}

.margin-top-3 {
  margin-top: 30px !important;
}

.margin-top-4 {
  margin-top: 40px !important;
}

.margin-bottom-0 {
  margin-bottom: 0 !important;
}

.margin-bottom-1 {
  margin-bottom: 8px !important;
}

.margin-bottom-2 {
  margin-bottom: 16px !important;
}

.margin-bottom-3 {
  margin-bottom: 30px !important;
}

.margin-bottom-4 {
  margin-bottom: 40px !important;
}

.margin-left-auto {
  margin-left: auto !important;
}

.margin-left-0 {
  margin-left: 0 !important;
}

.margin-left-1 {
  margin-left: 8px !important;
}

.margin-left-2 {
  margin-left: 16px !important;
}

.margin-right-auto {
  margin-right: auto !important;
}

.margin-right-0 {
  margin-right: 0 !important;
}

.margin-right-1 {
  margin-right: 8px !important;
}

.margin-right-2 {
  margin-right: 16px !important;
}

.resizing-horizontal {
  cursor: col-resize;
}

// This selector can only be used to temporarily show an element for calculating
// positions.
.forced-block {
  display: block !important;
}

// Sometimes we forcefully want to disable the the user-select property to initial
// because the element is for example being edited via a contenteditable attribute
// which does not work in combination with the user-select: none; property in Safari.
.forced-user-select-initial {
  user-select: initial !important;
}

.forced-text-overflow-initial {
  text-overflow: initial !important;
}

.prevent-scroll {
  overflow: hidden !important;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(720deg);
  }
}
