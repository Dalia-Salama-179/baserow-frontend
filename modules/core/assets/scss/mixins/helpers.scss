@mixin fixed-height($height, $font-size) {
  height: $height;
  line-height: $height;
  font-size: $font-size;
}

@mixin center-text($width, $font-size, $height: $width) {
  @include fixed-height($height, $font-size);

  width: $width;
  text-align: center;
}

@mixin loading($size: 1.4rem) {
  position: relative;
  display: block;
  width: $size;
  height: $size;
  border-radius: 50%;
  border: 0.25em solid;
  border-color: $color-primary-500 transparent $color-primary-500 transparent;
  animation: spin infinite 1800ms;
  animation-timing-function: cubic-bezier(0.785, 0.135, 0.15, 0.86);
}

// The basically works the same as padding: 10px 0 0 10px or padding: 10px; only then
// resulting position: absolute; with corresponding top, right, bottom, left values.
// -------
// Example:
//  @include absolute(10px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 10px; bottom: 10px; left: 10px;
// -------
// Example:
//  @include absolute(10px, 20px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 10px; left: 20px;
// -------
// Example:
//  @include absolute(10px, 20px, 30px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 30px;
@mixin absolute($top: null, $right: null, $bottom: null, $left: null) {
  position: absolute;

  @if ($top != null and $right == null and $bottom == null and $left == null) {
    top: $top;
    right: $top;
    bottom: $top;
    left: $top;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else if ($top != null and $right != null and $bottom == null and
  $left == null) {
    top: $top;
    right: $right;
    bottom: $top;
    left: $right;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else {
    @if $top != null {
      top: $top;
    }

    @if $right != null {
      right: $right;
    }

    @if $bottom != null {
      bottom: $bottom;
    }

    @if $left != null {
      left: $left;
    }
  }
}
