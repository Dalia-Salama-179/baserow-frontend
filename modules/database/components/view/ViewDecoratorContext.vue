<template>
  <Context>
    <ViewDecoratorList
      v-if="activeDecorations.length === 0"
      :view="view"
      @select="addDecoration($event)"
    />
    <div v-else class="decorator-context">
      <div class="decorator-context__list">
        <div
          v-for="dec in activeDecorations"
          :key="dec.decoration.id"
          class="decorator-context__decorator"
        >
          <div class="decorator-context__decorator-header">
            <div class="decorator-context__decorator-header-info">
              <ViewDecoratorItem :decorator-type="dec.decoratorType" />
            </div>
            <div
              v-show="dec.valueProviderType"
              class="decorator-context__decorator-header-select"
            >
              <Picker
                v-if="dec.valueProviderType"
                :icon="dec.valueProviderType.getIconClass()"
                :name="dec.valueProviderType.getName()"
                @select="selectValueProvider(dec.decoration, $event)"
              >
                <template #default="{ hidePicker }">
                  <div
                    class="
                      decorator-context__decorator-value-provider-select-list
                    "
                  >
                    <DecoratorValueProviderList
                      :decoration="dec.decoration"
                      @select="
                        ;[
                          selectValueProvider(dec.decoration, $event),
                          hidePicker(),
                        ]
                      "
                    />
                  </div>
                </template>
              </Picker>
            </div>
            <div class="decorator-context__decorator-header-trash">
              <a
                class="decorator-context__decorator-header-trash-link"
                @click="removeDecoration(dec.decoration)"
              >
                <i class="fa fa-trash" />
              </a>
            </div>
          </div>
          <component
            :is="dec.valueProviderType.getFormComponent()"
            v-if="dec.valueProviderType"
            :view="view"
            :table="table"
            :primary="primary"
            :fields="fields"
            :read-only="readOnly"
            :options="dec.decoration.value_provider_conf"
            @update="updateDecorationOptions(dec.decoration, $event)"
          />
          <DecoratorValueProviderList
            v-else
            :decoration="dec.decoration"
            :direction="'row'"
            :read-only="readOnly || dec.decoration._.loading"
            @select="selectValueProvider(dec.decoration, $event)"
          />
        </div>
      </div>
      <div class="decorator-context__footer">
        <a
          ref="addDecoratorLink"
          class="decorator-context__add"
          @click="
            $refs.selectDecoratorContext.toggle(
              $refs.addDecoratorLink,
              'bottom',
              'left',
              4
            )
          "
        >
          <i class="fas fa-plus" />
          {{ $t('viewDecoratorContext.addDecorator') }}
        </a>
        <SelectViewDecoratorContext
          ref="selectDecoratorContext"
          :view="view"
          @select="
            ;[$refs.selectDecoratorContext.hide(), addDecoration($event)]
          "
        />
      </div>
    </div>
  </Context>
</template>

<script>
import context from '@baserow/modules/core/mixins/context'
import viewDecoration from '@baserow/modules/database/mixins/viewDecoration'
import ViewDecoratorList from '@baserow/modules/database/components/view/ViewDecoratorList'
import ViewDecoratorItem from '@baserow/modules/database/components/view/ViewDecoratorItem'
import SelectViewDecoratorContext from '@baserow/modules/database/components/view/SelectViewDecoratorContext'
import DecoratorValueProviderItem from '@baserow/modules/database/components/view/DecoratorValueProviderItem'
import DecoratorValueProviderList from '@baserow/modules/database/components/view/DecoratorValueProviderList'
import { notifyIf } from '@baserow/modules/core/utils/error'

export default {
  name: 'ViewDecoratorContext',
  components: {
    ViewDecoratorList,
    ViewDecoratorItem,
    DecoratorValueProviderItem,
    SelectViewDecoratorContext,
    DecoratorValueProviderList,
  },
  mixins: [context, viewDecoration],
  props: {
    primary: {
      type: Object,
      required: true,
    },
    fields: {
      type: Array,
      required: true,
    },
    view: {
      type: Object,
      required: true,
    },
    table: {
      type: Object,
      required: true,
    },
    readOnly: {
      type: Boolean,
      required: true,
    },
  },
  methods: {
    async removeDecoration(decoration) {
      try {
        await this.$store.dispatch('view/deleteDecoration', {
          view: this.view,
          decoration,
        })
      } catch (error) {
        notifyIf(error, 'view')
      }
    },
    async selectValueProvider(decoration, valueProvider) {
      const valueProviderType = this.$registry.get(
        'decoratorValueProvider',
        valueProvider
      )
      try {
        await this.$store.dispatch('view/updateDecoration', {
          view: this.view,
          values: {
            value_provider_type: valueProviderType.getType(),
            value_provider_conf: valueProviderType.getDefaultConfiguration({
              view: this.view,
              fields: this.allTableFields,
            }),
          },
          decoration,
        })
      } catch (error) {
        notifyIf(error, 'view')
      }
    },
    async addDecoration(decoratorType) {
      const decoration = {
        type: decoratorType.getType(),
        value_provider_type: '',
        value_provider_conf: {},
      }
      try {
        await this.$store.dispatch('view/createDecoration', {
          view: this.view,
          values: decoration,
        })
      } catch (error) {
        notifyIf(error, 'view')
      }
    },
    async updateDecorationOptions(decoration, options) {
      try {
        await this.$store.dispatch('view/updateDecoration', {
          view: this.view,
          values: { value_provider_conf: options },
          decoration,
        })
      } catch (error) {
        notifyIf(error, 'view')
      }
    },
  },
}
</script>
