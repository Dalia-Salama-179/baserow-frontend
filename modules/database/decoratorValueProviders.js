import { Registerable } from '@baserow/modules/core/registry'

export class DecoratorValueProviderType extends Registerable {
  /**
   * A human readable name of the decorator provider type.
   */
  getName() {
    return null
  }

  /**
   * Returns a user description for this value provider.
   */
  getDescription() {}

  /**
   * The font awesome 5 icon name that is used as convenience for the user to
   * recognize certain valueProvider types. If you for example want the filter
   * icon, you must return 'filter' here. This will result in the classname
   * 'fas fa-filter'.
   */
  getIconClass() {
    return null
  }

  /**
   * Should return the view decorator type that the value provider is compatible with or
   * functions which take a view decoratorType and return a boolean indicating if the
   * value provider is compatible or not.
   *
   * You can also use a function like in this example:
   * [(decoratorType) => decoratorType.some_prop === 10, LeftBorderColorViewDecoratorType]
   * and then decoratorType which pass the test defined by the function will be
   * deemed as compatible.
   */
  getCompatibleDecoratorTypes() {
    return []
  }

  /**
   * Returns if a given view decoratorType is compatible with this value provider or
   * not.
   * Uses the list provided by `.getCompatibleDecoratorTypes()` to calculate this.
   */
  isCompatible(decorationType) {
    for (const typeOrFunc of this.getCompatibleDecoratorTypes()) {
      if (Object.prototype.hasOwnProperty.call(typeOrFunc, 'getType')) {
        if (decorationType.getType() === typeOrFunc.getType()) {
          return true
        }
      } else if (typeOrFunc instanceof Function) {
        if (typeOrFunc(decorationType)) {
          return true
        }
      }
    }
    return false
  }

  /**
   * Returns the component that allows the user to configure the value provider.
   * This component is responsible for creating the `value_provider_conf` object.
   *
   * If you want to reference fields in this object, you must use the key name
   * `field_id` to allow import/export to replace automatically any field id by the new
   * field id. `field_id`s can be anywhere in this object and this object can be as
   * deep as you need.
   */
  getFormComponent() {
    throw new Error(
      'Not implemented error. This value provider should return a component.'
    )
  }

  /**
   * Returns the default configuration when selecting this value provider from fields
   * and view
   * @returns the configuration object
   */
  getDefaultConfiguration({ fields, view }) {
    return {}
  }

  /**
   * Returns the value of this provider for the given row considering the configuration.
   *
   * @param {array} row the row
   * @param {object} options the configuration of the value provider
   * @param {array} fields the array of the fields of the current view
   *
   */
  getValue({ options, fields, row }) {
    throw new Error(
      'Not implemented error. This value provider should return a value.'
    )
  }

  getOrder() {
    return 50
  }

  /**
   * @return object
   */
  serialize() {
    return {
      type: this.type,
      name: this.getName(),
    }
  }
}
