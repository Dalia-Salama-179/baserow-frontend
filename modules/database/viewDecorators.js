import { Registerable } from '@baserow/modules/core/registry'

export class ViewDecoratorType extends Registerable {
  /**
   * A human readable name of the decorator type.
   */
  getName() {
    return null
  }

  /**
   * A description of the decorator type.
   */
  getDescription() {
    return null
  }

  /**
   * @returns the image URL to illustrate this decorator.
   */
  getImage() {
    return null
  }

  /**
   * If the decorator type is disabled, this text will be visible explaining why.
   */
  getDeactivatedText({ view }) {}

  /**
   * Indicates if the decorator type is disabled.
   */
  isDeactivated({ view }) {
    return false
  }

  /**
   * Returns whether or not the the user can add a new instance of this decorator.
   * A decorator might be disabled if, for example, there is already one occurrence
   * of the same type for the view.
   * The result must be an array. The first item is a boolean value, `true` if
   * the decorator is enabled. If not, it should be `false` and the second item
   * must be a user string describing the reason why it's not available.
   */
  canAdd({ view }) {
    return [false, '']
  }

  /**
   * Returns whether or not a given viewType is compatible with this view decorator.
   */
  isCompatible(view) {
    return false
  }

  /**
   * Should return the component that will actually decorate the record.
   */
  getComponent() {
    throw new Error(
      'Not implemented error. This view decorator should return a component.'
    )
  }

  /**
   * Returns the place where the decorator should appears. Allowed values are:
   * - `wrapper` if the decorator is a wrapper of the record.
   * - `first_cell` to decorate the first cell.
   */
  getPlace() {
    return null
  }

  getOrder() {
    return 50
  }

  /**
   * @return object
   */
  serialize() {
    return {
      type: this.type,
      name: this.getName(),
    }
  }
}
