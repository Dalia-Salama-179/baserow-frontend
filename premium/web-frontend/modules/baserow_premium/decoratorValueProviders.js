import { DecoratorValueProviderType } from '@baserow/modules/database/decoratorValueProviders'

import SingleSelectColorValueProviderForm from '@baserow_premium/components/views/SingleSelectColorValueProviderForm'
import ConditionalColorValueProviderForm from '@baserow_premium/components/views/ConditionalColorValueProviderForm'
import {
  BackgroundColorViewDecoratorType,
  LeftBorderColorViewDecoratorType,
} from '@baserow_premium/viewDecorators'
import { uuid } from '@baserow/modules/core/utils/string'
import { randomColor } from '@baserow/modules/core/utils/colors'
import { matchSearchFilters } from '@baserow/modules/database/utils/view'

export class SingleSelectColorValueProviderType extends DecoratorValueProviderType {
  static getType() {
    return 'single_select_color'
  }

  getName() {
    const { i18n } = this.app
    return i18n.t('decoratorValueProviderType.singleSelectColor')
  }

  getDescription() {
    const { i18n } = this.app
    return i18n.t('decoratorValueProviderType.singleSelectColorDescription')
  }

  getIconClass() {
    return 'chevron-circle-down'
  }

  getFormComponent() {
    return SingleSelectColorValueProviderForm
  }

  getCompatibleDecoratorTypes() {
    return [LeftBorderColorViewDecoratorType, BackgroundColorViewDecoratorType]
  }

  getValue({ options, fields, row }) {
    const value = row[`field_${options.field_id}`]
    return value?.color || ''
  }

  getDefaultConfiguration({ fields }) {
    const firstSelectField = fields.find(({ type }) => type === 'single_select')
    return {
      field_id: firstSelectField?.id || null,
    }
  }
}

export class ConditionalColorValueProviderType extends DecoratorValueProviderType {
  static getType() {
    return 'conditional_color'
  }

  static getDefaultFilterConf(registry, { fields }) {
    const field = fields[0]
    const filter = { field: field.id }

    const viewFilterTypes = registry.getAll('viewFilter')
    const compatibleType = Object.values(viewFilterTypes).find(
      (viewFilterType) => {
        return viewFilterType.fieldIsCompatible(field)
      }
    )

    filter.type = compatibleType.type
    const viewFilterType = registry.get('viewFilter', filter.type)
    filter.value = viewFilterType.getDefaultValue()
    filter.preload_values = {}
    filter.id = uuid()

    return filter
  }

  static getDefaultColorConf(registry, { fields }, noFilter = false) {
    return {
      color: randomColor(),
      operator: 'AND',
      filters: noFilter
        ? []
        : [
            ConditionalColorValueProviderType.getDefaultFilterConf(registry, {
              fields,
            }),
          ],
      uid: uuid(),
    }
  }

  getName() {
    const { i18n } = this.app
    return i18n.t('decoratorValueProviderType.conditionalColor')
  }

  getDescription() {
    const { i18n } = this.app
    return i18n.t('decoratorValueProviderType.conditionalColorDescription')
  }

  getIconClass() {
    return 'filter'
  }

  getValue({ options, fields, row }) {
    const { $registry } = this.app
    for (const { color, filters, operator } of options.colors) {
      if (
        row.id !== -1 &&
        row.id !== undefined &&
        matchSearchFilters($registry, operator, filters, fields, row)
      ) {
        return color
      }
    }
    return ''
  }

  getCompatibleDecoratorTypes() {
    return [LeftBorderColorViewDecoratorType, BackgroundColorViewDecoratorType]
  }

  getFormComponent() {
    return ConditionalColorValueProviderForm
  }

  getDefaultConfiguration({ fields }) {
    const { $registry } = this.app
    return {
      default: null,
      colors: [
        ConditionalColorValueProviderType.getDefaultColorConf($registry, {
          fields,
        }),
      ],
    }
  }
}
