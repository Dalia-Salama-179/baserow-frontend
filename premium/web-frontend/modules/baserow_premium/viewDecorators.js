import { ViewDecoratorType } from '@baserow/modules/database/viewDecorators'
import { PremiumPlugin } from '@baserow_premium/plugins'

import {
  GridViewType,
  GalleryViewType,
} from '@baserow/modules/database/viewTypes'

import { KanbanViewType } from './viewTypes'

import leftBorderDecoratorImage from '@baserow_premium/assets/images/leftBorderDecorator.svg'
import backgroundDecoratorImage from '@baserow_premium/assets/images/backgroundDecorator.svg'

import LeftBorderColorViewDecorator from '@baserow_premium/components/views/LeftBorderColorViewDecorator'
import BackgroundColorViewDecorator from '@baserow_premium/components/views/BackgroundColorViewDecorator'

export class LeftBorderColorViewDecoratorType extends ViewDecoratorType {
  static getType() {
    return 'left_border_color'
  }

  getName() {
    const { i18n } = this.app
    return i18n.t('viewDecoratorType.leftBorderColor')
  }

  getDescription() {
    const { i18n } = this.app
    return i18n.t('viewDecoratorType.leftBorderColorDescription')
  }

  getImage() {
    return leftBorderDecoratorImage
  }

  getPlace() {
    return 'first_cell'
  }

  getDeactivatedText() {
    const { i18n } = this.app
    return i18n.t('viewDecoratorType.onlyForPremium')
  }

  isDeactivated() {
    const { store } = this.app

    const additionalUserData = store.getters['auth/getAdditionalUserData']

    if (PremiumPlugin.hasValidPremiumLicense(additionalUserData)) {
      return false
    }
    return true
  }

  canAdd({ view }) {
    const { i18n } = this.app

    if (view.decorations.some(({ type }) => type === this.getType())) {
      return [false, i18n.t('viewDecoratorType.onlyOneDecoratorPerView')]
    }
    return [true]
  }

  getComponent() {
    const { store } = this.app

    // Read from the store
    const additionalUserData = store.getters['auth/getAdditionalUserData']

    if (PremiumPlugin.hasValidPremiumLicense(additionalUserData)) {
      return LeftBorderColorViewDecorator
    }

    return null
  }

  isCompatible(view) {
    const { store } = this.app

    return (
      [
        GridViewType.getType(),
        GalleryViewType.getType(),
        KanbanViewType.getType(),
      ].includes(view.type) && !store.getters['view/grid/isPublic']
    )
  }
}

export class BackgroundColorViewDecoratorType extends ViewDecoratorType {
  static getType() {
    return 'background_color'
  }

  getName() {
    const { i18n } = this.app
    return i18n.t('viewDecoratorType.backgroundColor')
  }

  getDescription() {
    const { i18n } = this.app
    return i18n.t('viewDecoratorType.backgroundColorDescription')
  }

  getImage() {
    return backgroundDecoratorImage
  }

  getPlace() {
    return 'wrapper'
  }

  getComponent() {
    const { store } = this.app

    // Read from the store
    const additionalUserData = store.getters['auth/getAdditionalUserData']

    if (PremiumPlugin.hasValidPremiumLicense(additionalUserData)) {
      return BackgroundColorViewDecorator
    }

    return null
  }

  isCompatible(view) {
    const { store } = this.app

    return (
      [
        GridViewType.getType(),
        GalleryViewType.getType(),
        KanbanViewType.getType(),
      ].includes(view.type) && !store.getters['view/grid/isPublic']
    )
  }

  getDeactivatedText() {
    const { i18n } = this.app
    return i18n.t('viewDecoratorType.onlyForPremium')
  }

  isDeactivated() {
    const { store } = this.app

    const additionalUserData = store.getters['auth/getAdditionalUserData']

    if (PremiumPlugin.hasValidPremiumLicense(additionalUserData)) {
      return false
    }
    return true
  }

  canAdd({ view }) {
    const { i18n } = this.app

    if (view.decorations.some(({ type }) => type === this.getType())) {
      return [false, i18n.t('viewDecoratorType.onlyOneDecoratorPerView')]
    }
    return [true]
  }
}
