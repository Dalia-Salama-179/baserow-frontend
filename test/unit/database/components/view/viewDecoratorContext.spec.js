import { TestApp } from '@baserow/test/helpers/testApp'
import ViewDecoratorContext from '@baserow/modules/database/components/view/ViewDecoratorContext'
import { DecoratorValueProviderType } from '@baserow/modules/database/decoratorValueProviders'
import { ViewDecoratorType } from '@baserow/modules/database/viewDecorators'

export class FakeDecoratorType extends ViewDecoratorType {
  static getType() {
    return 'fake_decorator'
  }

  getName() {
    return 'Fake decorator'
  }

  getDescription() {
    return 'Fake decorator description'
  }

  getImage() {
    return 'fake/url.png'
  }

  isCompatible(view) {
    return true
  }

  isDeactivated({ view }) {
    return false
  }

  getDeactivatedText() {
    return 'unavailability reason'
  }

  getComponent() {
    const component = {
      functional: true,

      render(h, ctx) {
        return h('div', `fake_decoration: ${ctx.props.value}`)
      },
    }
    return component
  }

  canAdd({ view }) {
    return [true]
  }
}

export class FakeValueProviderType extends DecoratorValueProviderType {
  static getType() {
    return 'fake_value_provider_type'
  }

  getName() {
    return 'Fake value provider'
  }

  getDescription() {
    return 'Fake value provider description.'
  }

  getIconClass() {
    return 'filter'
  }

  isCompatible() {
    return true
  }

  getFormComponent() {
    const component = {
      functional: true,

      render(h) {
        return h('div', 'fake_value_provider_form')
      },
    }
    return component
  }
}

const fieldData = [
  {
    id: 1,
    name: 'Name',
    order: 0,
    type: 'text',
    primary: true,
    text_default: '',
  },
  {
    id: 2,
    name: 'Surname',
    type: 'text',
    text_default: '',
    primary: false,
  },
  {
    id: 3,
    name: 'Status',
    type: 'single_select',
    text_default: '',
    primary: false,
  },
]

const rows = [
  { id: 2, order: '2.00000000000000000000', field_2: 'Bram' },
  { id: 4, order: '2.50000000000000000000', field_2: 'foo', field_3: 'bar' },
]

describe('GridViewRows component with decoration', () => {
  let testApp = null
  let mockServer = null
  let store = null
  let wrapperToDestroy = null

  beforeAll(() => {
    testApp = new TestApp()
    store = testApp.store
    mockServer = testApp.mockServer
    store.$registry.registerNamespace('viewDecorator')
    store.$registry.registerNamespace('decoratorValueProvider')
  })

  afterEach((done) => {
    // Clean up potentially registered stuff
    try {
      store.$registry.unregister('viewDecorator', 'fake_decorator')
    } catch {}
    try {
      store.$registry.unregister(
        'decoratorValueProvider',
        'fake_value_provider_type'
      )
    } catch {}
    if (wrapperToDestroy) {
      wrapperToDestroy.destroy()
      wrapperToDestroy = null
    }
    testApp.afterEach().then(done)
  })

  const populateStore = async ({ viewId = 1, decorations } = {}) => {
    const table = mockServer.createTable()
    const { application } = await mockServer.createAppAndGroup(table)
    const view = mockServer.createGridView(application, table, {
      decorations,
      viewId,
    })

    mockServer.createFields(application, table, fieldData)
    await store.dispatch('field/fetchAll', table)
    const primary = store.getters['field/getPrimary']
    const fields = store.getters['field/getAll']

    mockServer.createGridRows(view, fields, rows)
    await store.dispatch('page/view/grid/fetchInitial', {
      gridId: view.id,
      fields,
      primary,
    })
    await store.dispatch('view/fetchAll', { id: table.id })
    return { table, primary, fields, view }
  }

  const mountComponent = async (props) => {
    const wrapper = await testApp.mount(ViewDecoratorContext, {
      propsData: props,
    })

    await wrapper.vm.show(document.body)
    await wrapper.vm.$nextTick()

    // Allow to clean the DOM after the test
    wrapperToDestroy = wrapper

    return wrapper
  }

  test('Default component', async () => {
    const { table, primary, fields, view } = await populateStore()

    const fakeDecorator = new FakeDecoratorType({ app: testApp })
    const fakeValueProvider = new FakeValueProviderType({ app: testApp })

    store.$registry.register('viewDecorator', fakeDecorator)
    store.$registry.register('decoratorValueProvider', fakeValueProvider)

    const wrapper = await mountComponent({
      view,
      table,
      primary,
      fields,
      readOnly: false,
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('View with decoration configured', async () => {
    const { table, primary, fields, view } = await populateStore({
      decorations: [
        {
          type: 'fake_decorator',
          value_provider_type: 'fake_value_provider_type',
          value_provider_conf: {},
          _: { loading: false },
        },
        {
          type: 'fake_decorator',
          value_provider_type: '',
          value_provider_conf: {},
          _: { loading: false },
        },
      ],
    })

    const fakeDecorator = new FakeDecoratorType({ app: testApp })
    const fakeValueProvider = new FakeValueProviderType({ app: testApp })

    store.$registry.register('viewDecorator', fakeDecorator)
    store.$registry.register('decoratorValueProvider', fakeValueProvider)

    const wrapper = await mountComponent({
      view,
      table,
      primary,
      fields,
      readOnly: false,
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('Should show unavailable decorator tooltip', async () => {
    const { table, primary, fields, view } = await populateStore()

    const fakeDecorator = new FakeDecoratorType({ app: testApp })
    const fakeValueProvider = new FakeValueProviderType({ app: testApp })

    fakeDecorator.isDeactivated = () => true

    store.$registry.register('viewDecorator', fakeDecorator)
    store.$registry.register('decoratorValueProvider', fakeValueProvider)

    const wrapper = await mountComponent({
      view,
      table,
      primary,
      fields,
      readOnly: false,
    })

    await wrapper
      .find('.decorator-list > div:first-child')
      .trigger('mouseenter')

    expect(document.body).toMatchSnapshot()
  })

  test('Should show cant add decorator tooltip', async () => {
    const { table, primary, fields, view } = await populateStore()

    const fakeDecorator = new FakeDecoratorType({ app: testApp })
    const fakeValueProvider = new FakeValueProviderType({ app: testApp })

    fakeDecorator.canAdd = () => [false, "can't be added reason"]

    store.$registry.register('viewDecorator', fakeDecorator)
    store.$registry.register('decoratorValueProvider', fakeValueProvider)

    const wrapper = await mountComponent({
      view,
      table,
      primary,
      fields,
      readOnly: false,
    })

    await wrapper
      .find('.decorator-list > div:first-child')
      .trigger('mouseenter')

    expect(document.body).toMatchSnapshot()
  })
})
