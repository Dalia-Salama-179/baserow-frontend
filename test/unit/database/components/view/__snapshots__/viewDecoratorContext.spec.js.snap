// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridViewRows component with decoration Default component 1`] = `
<div
  class="context"
>
  <div
    class="decorator-list"
  >
    <div
      class="decorator-list__item"
    >
      <div
        class="decorator-item"
      >
        <img
          class="decorator-item__image"
          src="fake/url.png"
        />
         
        <div
          class="decorator-item__content"
        >
          <div
            class="decorator-item__title"
          >
            
      Fake decorator
    
          </div>
           
          <div
            class="decorator-item__description"
          >
            
      Fake decorator description
    
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GridViewRows component with decoration Should show cant add decorator tooltip 1`] = `
<body>
  <div
    class="tooltip tooltip--body tooltip--center"
    style="top: 4px; left: 0px;"
  >
    <div
      class="tooltip__content"
    >
      can't be added reason
    </div>
  </div>
  <div
    class="context"
  >
    <div
      class="decorator-list"
    >
      <div
        class="decorator-list__item decorator-list__item--disabled"
      >
        <div
          class="decorator-item"
        >
          <img
            class="decorator-item__image"
            src="fake/url.png"
          />
           
          <div
            class="decorator-item__content"
          >
            <div
              class="decorator-item__title"
            >
              
      Fake decorator
    
            </div>
             
            <div
              class="decorator-item__description"
            >
              
      Fake decorator description
    
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`GridViewRows component with decoration Should show unavailable decorator tooltip 1`] = `
<body>
  <div
    class="tooltip tooltip--body tooltip--center"
    style="top: 4px; left: 0px;"
  >
    <div
      class="tooltip__content"
    >
      unavailability reason
    </div>
  </div>
  <div
    class="context"
  >
    <div
      class="decorator-list"
    >
      <div
        class="decorator-list__item decorator-list__item--disabled"
      >
        <div
          class="decorator-item"
        >
          <img
            class="decorator-item__image"
            src="fake/url.png"
          />
           
          <div
            class="decorator-item__content"
          >
            <div
              class="decorator-item__title"
            >
              
      Fake decorator
    
            </div>
             
            <div
              class="decorator-item__description"
            >
              
      Fake decorator description
    
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`GridViewRows component with decoration View with decoration configured 1`] = `
<div
  class="context"
>
  <div
    class="decorator-context"
  >
    <div
      class="decorator-context__list"
    >
      <div
        class="decorator-context__decorator"
      >
        <div
          class="decorator-context__decorator-header"
        >
          <div
            class="decorator-context__decorator-header-info"
          >
            <div
              class="decorator-item"
            >
              <img
                class="decorator-item__image"
                src="fake/url.png"
              />
               
              <div
                class="decorator-item__content"
              >
                <div
                  class="decorator-item__title"
                >
                  
      Fake decorator
    
                </div>
                 
                <div
                  class="decorator-item__description"
                >
                  
      Fake decorator description
    
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="decorator-context__decorator-header-select"
          >
            <div
              class="dropdown"
            >
              <a
                class="dropdown__selected"
              >
                <i
                  class="dropdown__selected-icon fas fa-filter"
                />
                
    Fake value provider
    
                <i
                  class="dropdown__toggle-icon fas fa-caret-down"
                />
              </a>
               
            </div>
          </div>
           
          <div
            class="decorator-context__decorator-header-trash"
          >
            <a
              class="decorator-context__decorator-header-trash-link"
            >
              <i
                class="fa fa-trash"
              />
            </a>
          </div>
        </div>
         
        <div>
          fake_value_provider_form
        </div>
      </div>
      <div
        class="decorator-context__decorator"
      >
        <div
          class="decorator-context__decorator-header"
        >
          <div
            class="decorator-context__decorator-header-info"
          >
            <div
              class="decorator-item"
            >
              <img
                class="decorator-item__image"
                src="fake/url.png"
              />
               
              <div
                class="decorator-item__content"
              >
                <div
                  class="decorator-item__title"
                >
                  
      Fake decorator
    
                </div>
                 
                <div
                  class="decorator-item__description"
                >
                  
      Fake decorator description
    
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="decorator-context__decorator-header-select"
            style="display: none;"
          >
            <!---->
          </div>
           
          <div
            class="decorator-context__decorator-header-trash"
          >
            <a
              class="decorator-context__decorator-header-trash-link"
            >
              <i
                class="fa fa-trash"
              />
            </a>
          </div>
        </div>
         
        <div
          class="value-provider-list value-provider-list--row"
        >
          <div
            class="value-provider-list__item"
          >
            <div
              class="value-provider-item"
            >
              <div
                class="value-provider-item__title"
              >
                <i
                  class="fa fa-fw fa-filter value-provider-item__icon"
                />
                
    Fake value provider
  
              </div>
               
              <div
                class="value-provider-item__description"
              >
                
    Fake value provider description.
  
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="decorator-context__footer"
    >
      <a
        class="decorator-context__add"
      >
        <i
          class="fas fa-plus"
        />
        
        viewDecoratorContext.addDecorator
      
      </a>
       
    </div>
  </div>
</div>
`;
